desc/compute/_stability.py

```diff
@@ -231,3 +231,513 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+    provided in the list ``data`` above. Here's a description of
+    these parameters:
+
+    - a: minor radius of the device
+    - g^aa: |grad alpha|^2, field line bending term
+    - g^ra: (grad alpha dot grad rho) integrated local shear
+    - g^rr: |grad rho|^2 flux expansion term
+    - cvdrift: geometric factor of the curvature drift
+    - cvdrift0: geoetric factor of curvature drift 2
+    - |B|: magnitude of the magnetic field
+    - B^zeta: inverse of the jacobian
+    - p_r: dp/drho, pressure gradient
+    - phi: coordinate describing the position in the toroidal angle
+    along a field line
+    """
+    rho = data["rho"]
+
+    psi_b = params["Psi"] / (2 * jnp.pi)
+    a_N = data["a"]
+    B_N = 2 * psi_b / a_N**2
+
+    N_zeta0 = int(11)
+    # up-down symmetric equilibria only
+    zeta0 = jnp.linspace(-jnp.pi / 2, jnp.pi / 2, N_zeta0)
+
+    iota = data["iota"]
+    shat = -rho / iota * data["iota_r"]
+    psi = data["psi"]
+    sign_psi = jnp.sign(psi[-1])
+    sign_iota = jnp.sign(iota[-1])
+
+    phi = data["phi"]
+
+    N_alpha = int(8)
+    N_zeta = int(len(phi) / N_alpha)
+
+    B = jnp.reshape(data["|B|"], (N_alpha, 1, N_zeta))
+    gradpar = jnp.reshape(data["B^zeta"] / data["|B|"], (N_alpha, 1, N_zeta))
+    dpdpsi = jnp.mean(mu_0 * data["p_r"] / data["psi_r"])
+
+    gds2 = jnp.reshape(
+        rho**2
+        * (
+            data["g^aa"][None, :]
+            - 2 * sign_iota * shat / rho * zeta0[:, None] * data["g^ra"][None, :]
+            + zeta0[:, None] ** 2 * (shat / rho) ** 2 * data["g^rr"][None, :]
+        ),
+        (N_alpha, N_zeta0, N_zeta),
+    )
+
+    f = a_N**3 * B_N * gds2 / B**3 * 1 / gradpar
+    g = a_N**3 * B_N * gds2 / B * gradpar
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c = (
+        1
+        * a_N**3
+        * B_N
+        * jnp.reshape(
+            2
+            / data["B^zeta"][None, :]
+            * sign_psi
+            * rho**2
+            * dpdpsi
+            * (
+                data["cvdrift"][None, :]
+                - shat / rho * zeta0[:, None] * data["cvdrift0"][None, :]
+            ),
+            (N_alpha, N_zeta0, N_zeta),
+        )
+    )
+
+    h = phi[1] - phi[0]
+
+    A = jnp.zeros((N_alpha, N_zeta0, N_zeta - 2, N_zeta - 2))
+
+    i = jnp.arange(N_alpha)[:, None, None, None]
+    l = jnp.arange(N_zeta0)[None, :, None, None]
+    j = jnp.arange(N_zeta - 2)[None, None, :, None]
+    k = jnp.arange(N_zeta - 2)[None, None, None, :]
+
+    A = A.at[i, l, j, k].set(
+        g_half[i, l, k] / f[i, l, k] * 1 / h**2 * (j - k == -1)
+        + (
+            -(g_half[i, l, j + 1] + g_half[i, l, j]) / f[i, l, j + 1] * 1 / h**2
+            + c[i, l, j + 1] / f[i, l, j + 1]
+        )
+        * (j - k == 0)
+        + g_half[i, l, j] / f[i, l, j + 1] * 1 / h**2 * (j - k == 1)
+    )
+
+    w = eigvals(jnp.where(jnp.isfinite(A), A, 0))
+
+    lam = jnp.real(jnp.max(w, axis=(2,)))
+
+    lam = jnp.max(lam)
```

Comment:
do we want to only return the max? might be better to return all of them and only do the max in the objective function? (i could also see using softmax or something, or penalizing all of the positive growth rates, not just the max?)

Reply:
Ok, I'll change this